runtime/sema/check_binary_expression.go

```diff
@@ -293,18 +297,31 @@ func (checker *Checker) checkBinaryExpressionArithmeticOrNonEqualityComparisonOr
 		)
 	}
 
-	switch operationKind {
-	case BinaryOperationKindArithmetic,
-		BinaryOperationKindBitwise:
-
-		return leftType
+	return leftType
+}
 
-	case BinaryOperationKindNonEqualityComparison:
-		return BoolType
+func (checker *Checker) checkBinaryExpressionNonEquality(
+	expression *ast.BinaryExpression,
+	operation ast.Operation,
+	leftType, rightType Type,
+	anyInvalid bool,
+) (resultType Type) {
+	resultType = BoolType
 
-	default:
-		panic(errors.NewUnreachableError())
+	if !(leftType.Equal(rightType) && leftType.IsComparable() && rightType.IsComparable()) {
+		if !anyInvalid {
+			checker.report(
+				&InvalidBinaryOperandsError{
+					Operation: operation,
+					LeftType:  leftType,
+					RightType: rightType,
+					Range:     ast.NewRangeFromPositioned(checker.memoryGauge, expression),
+				},
+			)
+		}
 	}
```

Comment:
Simplify by unnesting

```suggestion
	if !(leftType.Equal(rightType) && leftType.IsComparable() && rightType.IsComparable()) && !anyInvalid {
		checker.report(
			&InvalidBinaryOperandsError{
				Operation: operation,
				LeftType:  leftType,
				RightType: rightType,
				Range:     ast.NewRangeFromPositioned(checker.memoryGauge, expression),
			},
		)
	}
```